deepali/losses/functional.py

```diff
@@ -636,6 +639,83 @@ def ssd_loss(
     return loss
 
 
+def mi_loss(
+        input: Tensor,
+        target: Tensor,
+        mask: Optional[Tensor] = None,
+        vmin: Optional[float] = None,
+        vmax: Optional[float] = None,
+        num_bins: int = 64,
+        sample_ratio: Optional[float] = None,
+        nmi: bool = True
+) -> Tensor:
+    r"""
+    Calculate mutual information loss using Parzen window density and entropy estimations
+
+    References:
+    - Qiu, H., Qin, C., Schuh, A., Hammernik, K.: Learning Diffeomorphic and Modality-invariant Registration using B-splines. Medical Imaging with Deep Learning. (2021).
+    - Thévenaz, P., Unser, M.: Optimization of mutual information for multiresolution image registration. IEEE Trans. Image Process. 9, 2083–2099 (2000).
+
+    Args:
+        input: Source image sampled on ``target`` grid.
+        target: Target image with same shape as ``input``.
+        mask: Multiplicative mask with same shape as ``input``.
+        vmin: Minimal intensity value the joint and marginal density is estimated.
+        vmax: Maximal intensity value the joint and marginal density is estimated.
+        num_bins: Number of bin edges to discretize the density estimation.
+        sample_ratio: Ratio of the voxels to randomly sample from the images,
+            if `None` mutual information is calculated using all voxels .
+        nmi: Calculate Normalised Mutual Information instead of Mutual Information if True.
+    """
+    input = masked_loss(input, mask=mask)
```

Comment:
Wouldn't it be more appropriate to ensure that samples from which the intensity distribution is estimated only include voxels within the mask?

Reply:
It would be more appropriate in theory. But if we have foreground masks with varying numbers of foreground voxels in a batch, sampling the same ratio of voxels means the sizes along the batch dimension would not be the same, which is not allowed in the tensor structure. 

there might be some hack around this issue within a batch. But if the tensor size changes between batches, it might lead to memory fragmentation issues in deep learning training, leading to higher GPU memory usage.

Reply:
I can think of a perhaps naive hack around both these issues, which is to use the bounding box of the foreground mask padded to the same pre-defined size. Then we sample within this bounding box. But this seems a bit too hacky?

Reply:
Do you think `torch.randperm` could be replaced with `torch.multinomial(mask)` to achieve this?

That function is limited, however, by the size of the mask which cannot exceed 2 ** 24 grid element. Hence, we would need to adapt a custom implementation of inverse transform sampling. See below snippet copied from some other code where I sample random image patches for example. You might want to draw without replacement.

```
        # Indices of randomly sampled image patch ROI centers
        #
        # It is important to draw samples with replacement here. Otherwise, smaller ROIs
        # such as in particular diseased regions are undersampled even if their respective sampling
        # weight is considerably higher. Duplicate samples may be discarded afterwards again based on
        # the maximum ROI IoU. For this, we draw more samples than required (cf. num_samples above).
        #
        # As torch.multinomial() works only for at most 2^24 categories, use inverse transform sampling
        # if the number of candidate voxels with a sampling weight greater than zero is very large.
        if len(candidates) > 2 ** 24:
            weight_cdf = weight.astype(np.float64).flatten()
            weight_cdf = np.cumsum(weight_cdf / weight_cdf.sum())
            weight_cdf = np.divide(weight_cdf, weight_cdf[-1], out=weight_cdf)
            cdf_values = torch.rand(num_samples, generator=self.generator)
            imroi_index = np.searchsorted(weight_cdf, cdf_values.numpy())
            imroi_index = np.clip(imroi_index, 0, weight_cdf.size - 1, out=imroi_index)
        else:
            w = torch.from_numpy(weight.flatten()[candidates]).type(torch.float32)
            i = torch.multinomial(w, num_samples, generator=self.generator, replacement=True)
            imroi_index = candidates[i.numpy()]
```